activerecord/lib/active_record/database_configurations/database_config.rb

```diff
@@ -6,11 +6,60 @@ class DatabaseConfigurations
     # UrlConfig respectively. It will never return a DatabaseConfig object,
     # as this is the parent class for the types of database configuration objects.
     class DatabaseConfig # :nodoc:
+      include Mutex_m
+
       attr_reader :env_name, :spec_name
 
+      attr_accessor :schema_cache
+
+      INSTANCES = ObjectSpace::WeakMap.new
+      private_constant :INSTANCES
+
+      class << self
+        def discard_pools!
+          INSTANCES.each_key(&:discard_pool!)
+        end
+      end
+
       def initialize(env_name, spec_name)
+        super()
         @env_name = env_name
         @spec_name = spec_name
+        @pool = nil
+
+        INSTANCES[self] = self
+      end
+
+      def disconnect!
+        ActiveSupport::ForkTracker.check!
+
+        return unless @pool
+
+        synchronize do
+          return unless @pool
+
+          @pool.automatic_reconnect = false
+          @pool.disconnect!
+        end
+
+        nil
+      end
+
+      def connection_pool
+        ActiveSupport::ForkTracker.check!
+
+        @pool || synchronize { @pool ||= ConnectionAdapters::ConnectionPool.new(self) }
+      end
+
+      def discard_pool!
+        return unless @pool
+
+        synchronize do
+          return unless @pool
+
+          @pool.discard!
+          @pool = nil
+        end
```

Comment:
Hey I totally missed this before so that's my bad, but I don't think this belongs in DatabaseConfig. 

DatabaseConfig should only be responsible for taking the yaml/provided config and turning it into objects that we then pass around in the pool and whatnot. It shouldn't be responsible for knowing anything about the connection_pool or how to discard the pool. I think this is starting to mix concerns a bit too much.

Reply:
I believe we need a object that does what this object is doing right now. It is a aggregation of the database configuration state. It does know how to answer that is the connection pool for that connection, what is the schema cache for that connection and how is the connection configured.

It doesn't make sense to me for the connection pool to own the information about the schema cache and the configuration.

I agree we don't need to reuse an existent object to do that.

Reply:
@eileencodes Totally agree that there's two mixed concerns here.

These two concerns will most definitely be split again once I introduce the topology classes/objects, but for now this made the calling code much simpler.

Reply:
👋 Just to add a bit here, we discovered this while working through a test failure on https://github.com/rails/rails/pull/37368 and it definitely makes it harder to pass around and re-use these config objects when they have state on them like `@pool`.

Thank you for your work on this! ✨ 

We can help move this around too if you'd like or don't think you'll get to it

Reply:
Well, not everything is in place yet, but ultimately I want a "role"  class such as:

```ruby
class Role
  @db_config
  @pool 
```

But it should be simpler to introduce after https://github.com/rails/rails/pull/37388.